File: binding/src/libqmimedatabase.cpp
#include <QByteArray>
#include <QFileInfo>
#include <QIODevice>
#include <QMimeDatabase>
#include <QMimeType>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QUrl>
#include <qmimedatabase.h>
#include "libqmimedatabase.h"
#include "libqmimedatabase.hxx"

QMimeDatabase* QMimeDatabase_new() {
	 return new QMimeDatabase();
}

QMimeType* QMimeDatabase_MimeTypeForName(const QMimeDatabase* self, const libqt_string nameOrAlias) {
	return new QMimeType(self->mimeTypeForName(QString::fromUtf8(nameOrAlias.data, nameOrAlias.len)));
}

QMimeType* QMimeDatabase_MimeTypeForFile(const QMimeDatabase* self, const libqt_string fileName) {
	return new QMimeType(self->mimeTypeForFile(QString::fromUtf8(fileName.data, fileName.len)));
}

QMimeType* QMimeDatabase_MimeTypeForFile2(const QMimeDatabase* self, const QFileInfo* fileInfo) {
	return new QMimeType(self->mimeTypeForFile(*fileInfo));
}

libqt_list QMimeDatabase_MimeTypesForFileName(const QMimeDatabase* self, const libqt_string fileName) {
	auto _ret = self->mimeTypesForFileName(QString::fromUtf8(fileName.data, fileName.len));
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

QMimeType* QMimeDatabase_MimeTypeForData(const QMimeDatabase* self, const libqt_string data) {
	return new QMimeType(self->mimeTypeForData(QByteArray(data.data, data.len)));
}

QMimeType* QMimeDatabase_MimeTypeForData2(const QMimeDatabase* self, QIODevice* device) {
	return new QMimeType(self->mimeTypeForData(device));
}

QMimeType* QMimeDatabase_MimeTypeForUrl(const QMimeDatabase* self, const QUrl* url) {
	return new QMimeType(self->mimeTypeForUrl(*url));
}

QMimeType* QMimeDatabase_MimeTypeForFileNameAndData(const QMimeDatabase* self, const libqt_string fileName, QIODevice* device) {
	return new QMimeType(self->mimeTypeForFileNameAndData(QString::fromUtf8(fileName.data, fileName.len), device));
}

QMimeType* QMimeDatabase_MimeTypeForFileNameAndData2(const QMimeDatabase* self, const libqt_string fileName, const libqt_string data) {
	return new QMimeType(self->mimeTypeForFileNameAndData(QString::fromUtf8(fileName.data, fileName.len), QByteArray(data.data, data.len)));
}

libqt_string QMimeDatabase_SuffixForFileName(const QMimeDatabase* self, const libqt_string fileName) {
	QString _ret = self->suffixForFileName(QString::fromUtf8(fileName.data, fileName.len));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_list QMimeDatabase_AllMimeTypes(const QMimeDatabase* self) {
	auto _ret = self->allMimeTypes();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

QMimeType* QMimeDatabase_MimeTypeForFile22(const QMimeDatabase* self, const libqt_string fileName, int mode) {
	return new QMimeType(self->mimeTypeForFile(QString::fromUtf8(fileName.data, fileName.len), static_cast<QMimeDatabase::MatchMode>(mode)));
}

QMimeType* QMimeDatabase_MimeTypeForFile23(const QMimeDatabase* self, const QFileInfo* fileInfo, int mode) {
	return new QMimeType(self->mimeTypeForFile(*fileInfo, static_cast<QMimeDatabase::MatchMode>(mode)));
}

void QMimeDatabase_Delete(QMimeDatabase* self) {
    delete self;
}

File: binding/include/libqmimedatabase.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qmimedatabase.html
pub const qmimedatabase = struct {

    /// New constructs a new QMimeDatabase object.
    pub fn New() QtC.QMimeDatabase {
        return qtc.QMimeDatabase_new();
    }


    pub fn MimeTypeForName(self: ?*anyopaque, nameOrAlias: []const u8) QtC.QMimeType {
        const nameOrAlias_str = qtc.libqt_string{
    .len = nameOrAlias.len,
    .data = nameOrAlias.ptr,
};
return qtc.QMimeDatabase_MimeTypeForName(@ptrCast(self), nameOrAlias_str);
    }

    pub fn MimeTypeForFile(self: ?*anyopaque, fileName: []const u8) QtC.QMimeType {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QMimeDatabase_MimeTypeForFile(@ptrCast(self), fileName_str);
    }

    pub fn MimeTypeForFile2(self: ?*anyopaque, fileInfo: ?*anyopaque) QtC.QMimeType {
        return qtc.QMimeDatabase_MimeTypeForFile2(@ptrCast(self), @ptrCast(fileInfo));
    }

    pub fn MimeTypesForFileName(self: ?*anyopaque, fileName: []const u8) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QMimeDatabase_MimeTypesForFileName(@ptrCast(self), fileName_str);
    }

    pub fn MimeTypeForData(self: ?*anyopaque, data: []u8) QtC.QMimeType {
        const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QMimeDatabase_MimeTypeForData(@ptrCast(self), data_str);
    }

    pub fn MimeTypeForData2(self: ?*anyopaque, device: ?*anyopaque) QtC.QMimeType {
        return qtc.QMimeDatabase_MimeTypeForData2(@ptrCast(self), @ptrCast(device));
    }

    pub fn MimeTypeForUrl(self: ?*anyopaque, url: ?*anyopaque) QtC.QMimeType {
        return qtc.QMimeDatabase_MimeTypeForUrl(@ptrCast(self), @ptrCast(url));
    }

    pub fn MimeTypeForFileNameAndData(self: ?*anyopaque, fileName: []const u8, device: ?*anyopaque) QtC.QMimeType {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QMimeDatabase_MimeTypeForFileNameAndData(@ptrCast(self), fileName_str, @ptrCast(device));
    }

    pub fn MimeTypeForFileNameAndData2(self: ?*anyopaque, fileName: []const u8, data: []u8) QtC.QMimeType {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const data_str = qtc.libqt_string{
    .len = data.len,
    .data = data.ptr,
};
return qtc.QMimeDatabase_MimeTypeForFileNameAndData2(@ptrCast(self), fileName_str, data_str);
    }

    pub fn SuffixForFileName(self: ?*anyopaque, fileName: []const u8, allocator: std.mem.Allocator) []const u8 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
const _str = qtc.QMimeDatabase_SuffixForFileName(@ptrCast(self), fileName_str);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qmimedatabase.SuffixForFileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AllMimeTypes(self: ?*anyopaque, ) []const u8 {
        return qtc.QMimeDatabase_AllMimeTypes(@ptrCast(self));
    }

    pub fn MimeTypeForFile22(self: ?*anyopaque, fileName: []const u8, mode: i32) QtC.QMimeType {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QMimeDatabase_MimeTypeForFile22(@ptrCast(self), fileName_str, @intCast(mode));
    }

    pub fn MimeTypeForFile23(self: ?*anyopaque, fileInfo: ?*anyopaque, mode: i32) QtC.QMimeType {
        return qtc.QMimeDatabase_MimeTypeForFile23(@ptrCast(self), @ptrCast(fileInfo), @intCast(mode));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QMimeDatabase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/mimedatabase.html#types
pub const enums = struct {
    pub const MatchMode = enum {
        pub const MatchDefault: i32 = 0;
        pub const MatchExtension: i32 = 1;
        pub const MatchContent: i32 = 2;
    };

};
